rfd/0147-repo-mirror.md

```diff
@@ -0,0 +1,73 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+
+# RFD 144 - Automatic Repo Mirroring
+
+## What
+
+This RFD outlines a method for continuously and automatically syncing a GitHub repository. This can be used to create private copies of a repository.
+
+## Why
+
+We currently face a few frictions which could be addressed through a more robust repository sync mechanism:
+1. Our internal `teleport-private` mirror of `teleport`, used for security development, currently requires manual updates to sync.
+2. Certain tools, like Dependabot, don't support scanning non-default branches, necessitating manual efforts to manage our release branches.
+3. Due to the committing of CI and scanning configurations in the repository, we receive duplicate alerts when using a repository for testing or as an internal mirror/fork.
+
+While addressing these immediate needs, the aim of this RFD is to establish a generic pattern for repository mirroring. This will provide flexibility for future requirements when syncing any upstream repository to a fork or mirror that we maintain.
+
+## Details
+
+The core requirements of this solution are:
+* Automatic and regular updates to a repository to ensure is's state matches the upstream.
+* The ability to make additional changes to the mirror repository, separate from the upstream repository (e.g., disabling actions we don't want to run on the mirror, or enabling additional scanning/actions).
+
+### Branch Structure
+
+* `master` (configurable) - This will serve as the repository's default branch. It will contain both the upstream history and any commits necessary for the synchronization process or other custom changes made to the mirror repository.
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch. Should there be a need to create a PR in this repository, the history should be based on this branch, not on `master`.
+* `sync/rebase` - This is the branch where custom changes will be committed. It will be rebased onto the upstream changes before being merged into master.
+
+#### Branch Protections
+
+The only protected branch will be `sync/rebase`. Other branches will allow force pushes to enable forceful synchronization based on the upstream state.
```

Comment:
Sorry for touching this again, but it still seems off to me that `master` isn't protected. Do we intend to let tooling force-push to it? Couldn't we use a non-squashed merge commit for the syncing?

For example, this is what we do for go-libfido2: https://gravitational.slab.com/posts/web-authn-and-passwordless-development-avzq9xn1#h0m6u-how-to-update-gravitational-go-libfido-2.

Reply:
Forced pushing allows us to keep the history as uniform as possible, and also highlights the custom changes we have applied at the top of the history.  The `master` of these repos should not be depended on for anything besides scanning and non-critical automation.

Doing a non-squashed merge commit is a possible option, but I am concerned it will create a confusing commit history that will be more likely to conflict.

Reply:
Using merge commits you can, with care, keep upstream commit hashes identical. Take a look at the histories below:

* https://github.com/gravitational/go-libfido2/commits/master
* https://github.com/gravitational/go-libfido2/commits/teleport

That said, in this model, our changes eventually get "buried" in the history while with rebases they'll always be on top.

(Feel free to resolve this thread.)

Reply:
You make a good point which I also just commented on here: https://github.com/gravitational/teleport/pull/32233#discussion_r1339029673

Although this strategy can keep a fork branch in sync with an external upstream, this may not be the best tool for that use case.
* I agree we don't want to rewrite history we actually are depending on, so a rebase and force push is not a good model for these forks
* I also am unsure if we want to pull in upstream changes without review

That said it would only merge changes which don't conflict, and we still need to review the inclusion of an update in `teleport`.

Maybe expanding the scope to all forks makes this more difficult than I originally thought, or maybe it's just as easy as supporting both methods.

Reply:
> Maybe expanding the scope to all forks makes this more difficult

Yeah, I think this is the trap I fell into here. For "readonly" mirrors with the occasional proposed change this is fine, for forky-forks not so much.

Reply:
I clarified why this process is not a good match for external forks (at least not yet): d937a7f999580eacfe6ce9eb174964240f11f39f